Add rudimentary mypy type checking

bluetech opened the pull request:
(This PR is currently an RFC, I will discuss it in issue #3342).

Add a very lax mypy configuration, add it to tox -e linting, and
fix/ignore the few errors that come up. The idea is to get it running
before diving in too much.

This enables:

- Progressively adding type annotations and enabling more strict
  options, which will improve the codebase (IMO).

- Annotating the public API in-line, and eventually exposing it to
  library users who use type checkers (with a py.typed file).

Though, none of this is done yet.

Refs #3342.

asottile commented on tox.ini:
```diff
@@ -53,8 +53,12 @@ platform = {env:_PYTEST_TOX_PLATFORM:.*}
+    mypy==0.711
+commands =
+    pre-commit run --all-files --show-diff-on-failure
+    mypy bench/ extra/ scripts/ src/ testing/
```
you can use mypy through pre-commit: https://github.com/pre-commit/mirrors-mypy

bluetech replied:
I tried it initially, but couldn't get it to work. It errors with a lot of module name conflicts, like someone reported here: https://github.com/python/mypy/issues/4008.

It also says it runs in an isolated venv which means 3rd-party stubs can't be used.

Maybe it can be improved to fix these things?

asottile replied:
do you have specifics for the first? we use this at work and it works fine there
the second is solved using `additional_dependencies` -- no different from putting them in `deps` in `tox.ini`

bluetech replied:
> do you have specifics for the first? we use this at work and it works fine there

Using this diff:

```diff
diff --git a/.pre-commit-config.yaml b/.pre-commit-config.yaml
index 0ff593289..4a53d7582 100644
--- a/.pre-commit-config.yaml
+++ b/.pre-commit-config.yaml
@@ -43,6 +43,10 @@ repos:
     rev: v1.4.0
     hooks:
     -   id: rst-backticks
+-   repo: https://github.com/pre-commit/mirrors-mypy
+    rev: v0.711
+    hooks:
+    -   id: mypy
 -   repo: local
     hooks:
     -   id: rst
```

mypy fails with errors like these:

```
doc/en/example/assertion/global_testmodule_config/conftest.py: error: Duplicate module named 'conftest' (also at 'testing/example_scripts/config/collect_pytest_prefix/conftest.py')
doc/en/example/assertion/global_testmodule_config/conftest.py: error: Are you missing an __init__.py?
```

Once you "fix" it (i.e. try to delete), another pops up.

>  the second is solved using additional_dependencies -- no different from putting them in deps in tox.ini

Right, it's the same. Though how easy would it be to get it to install the correct versions without repeating them?

Another thing is that it doesn't seem to pick up the configurations in `setup.cfg`. Can maybe use a `mypy.ini` and make it point to it? (I used a config file rather than flags so that code editors like vscode and manual `mypy` runs can pick them up automatically).

asottile replied:
> mypy fails with errors like these

ah yes that's PEP 420, if you add `__init__.py` files then mypy will recognize those as separate modules

> Right, it's the same. Though how easy would it be to get it to install the correct versions without repeating them?

you already have to repeat them in `tox` so it's no different

> Another thing is that it doesn't seem to pick up the configurations in `setup.cfg`

could you elaborate on "not respecting `setup.cfg`?" -- it's probably the same issue (`mypy-x.y.z` only works if there's `__init__.py` files in `x/`, `x/y`)

asottile replied:
the other thing you can do is split up the mypy invocations:

```yaml
    -   id: mypy
        name: mypy (src)
        files: ^src/
    -   id: mypy
        name: mypy (testing)
        files: ^testing/
    # ...
```

that way it doesn't try and dupe the modules

asottile commented on src/_pytest/tmpdir.py:
```diff
@@ -26,7 +26,9 @@ class TempPathFactory:
         # Path.absolute() exists, but it is not public (see https://bugs.python.org/issue25012)
-        converter=attr.converters.optional(lambda p: Path(os.path.abspath(str(p))))
+        converter=attr.converters.optional(
+            lambda p: Path(os.path.abspath(str(p)))  # type: ignore
```
what's the type error here?

bluetech replied:
I'll add more extensive code comments explaining each `type: ignore`.

asottile commented on src/_pytest/python.py:
```diff
@@ -1356,7 +1356,9 @@ def write_docstring(tw, doc, indent="    "):
 
 
-class Function(FunctionMixin, nodes.Item, fixtures.FuncargnamesCompatAttr):
+class Function(  # type: ignore
```
what's the ignore here?

asottile commented on src/_pytest/mark/structures.py:
```diff
@@ -298,7 +299,7 @@ def test_function():
 
     _config = None
-    _markers = set()
+    _markers: Set[str] = set()
```
this is python3.6+ syntax, we can't use this yet (still support python 3.5) -- you'll need type comments here

bluetech replied:
Fixed, added `python_version = 3.5` which catches these things.

asottile replied:
assumedly our testsuite should catch that as well so you shouldn't need that setting :thinking:

asottile commented on src/_pytest/fixtures.py:
```diff
@@ -960,7 +962,7 @@ class FixtureFunctionMarker:
     params = attr.ib(converter=attr.converters.optional(tuple))
     autouse = attr.ib(default=False)
-    ids = attr.ib(default=None, converter=_ensure_immutable_ids)
+    ids = attr.ib(default=None, converter=_ensure_immutable_ids)  # type: ignore
```
what's the type error here?

asottile commented on src/_pytest/debugging.py:
```diff
@@ -74,7 +74,7 @@ class pytestPDB:
     _pluginmanager = None
     _config = None
-    _saved = []
+    _saved: list = []
```
this isn't a useful annotation -- when mypy says "need annotation for ..." it's requesting that you fill out the generic parameter (it already knows it is a list, but "what list?")

bluetech replied:
Agreed, but I wasn't sure about this one -- currently I just want to get it running, so did the quick fix. When we start rejecting `Any` we'll need to confront this.

asottile commented on src/_pytest/fixtures.py:
```diff
@@ -54,10 +56,10 @@ def pytest_sessionstart(session):
 
 
-scopename2class = {}
+scopename2class: Dict[str, type] = {}
```
this should be `Dict[str, Type[baseclass]` where `baseclass` is a common ancestor of the types used here -- (iirc it's `Node`? I'd have to look it up can't tell directly from the diff :D)

asottile commented on src/_pytest/capture.py:
```diff
@@ -618,7 +618,7 @@ class FDCapture(FDCaptureBinary):
     """
 
-    EMPTY_BUFFER = str()
+    EMPTY_BUFFER = str()  # type: ignore
```
what's the type error here?  this should also probably just be `''` instead of `str()` -- maybe I'll teach `pyupgrade` to do that :thinking:  https://github.com/asottile/pyupgrade/issues/187

bluetech replied:
There superclass `FDCaptureBinary` sets this to `bytes`. I didn't try to understand too closely what this does TBH.

asottile replied:
`FDCapture` should probably be generic

asottile commented on src/_pytest/capture.py:
```diff
@@ -585,7 +585,7 @@ def snap(self):
         return res
 
-    def done(self):
+    def done(self):  # type: ignore
```
what's the type error here?

asottile commented on src/_pytest/capture.py:
```diff
@@ -568,7 +568,7 @@ def __repr__(self):
         )
 
-    def start(self):
+    def start(self):  # type: ignore
```
what's the type error here?

asottile replied:
I'd rather see `assert` guards for lateinit things -- that way we know in our tests when the assumption is violated and mypy can continue to typecheck the rest of the function -- I generally reserve `cast` / `# type: ignore` for actual mypy bugs only

bluetech replied:
Not sure what you mean about `assert` guards, however I'll add a different workaround for this, to avoid the clash between a method name and the `__init__` initialization.

asottile replied:
for example:

```python
class C:
    def __init__(self) -> None:
        self.x = None  # type: Optional[int]

    def set_x(self, x: int) -> None:
        self.x = x

    def use_x(self) -> int:
        assert self.x is not None  # convince mypy that lateinit has occurred
        return self.x + 5
```

without the assertion:

```console
$ mypy t.py
t.py:13: error: Unsupported operand types for + ("None" and "int")
t.py:13: note: Left operand is of type "Optional[int]"
```

with the assert:

```console
$ mypy t.py
$
```

bluetech replied:
Ah, right. The problem here is different ("foo is already defined").

asottile replied:
oh yikes! https://github.com/pytest-dev/pytest/blob/73767268eca69bb07bb71dfb4ab7b651295e0106/src/_pytest/capture.py#L546-L548

asottile commented on src/_pytest/assertion/rewrite.py:
```diff
@@ -459,39 +463,40 @@ def _fix(node, lineno, col_offset):
     tokens = tokenize.tokenize(io.BytesIO(src).readline)
-    for tp, src, (lineno, offset), _, line in tokens:
-        if tp == tokenize.NAME and src == "assert":
+    for tp, string, (lineno, offset), _, line in tokens:
```
`string` shadows the module by the same name -- perhaps choose a different name here? -- smaller diff would be to rename the parameter to `source` instead

asottile commented on src/_pytest/_code/code.py:
```diff
@@ -255,10 +256,10 @@ def __str__(self):
+    def _name_get(self):
         return self.frame.code.raw.co_name
 
-    name = property(name, None, None, "co_name of underlaying code")
```
this looks like python2.4 compatibility code -- can probably just use `@property` syntax

asottile replied:
this one too :)

asottile commented on setup.cfg:
```diff
@@ -61,3 +61,25 @@ ignore =
+# disallow_untyped_calls = True
+# disallow_untyped_decorators = True
+# disallow_untyped_defs = True
+ignore_missing_imports = True
```
when run through `pre-commit` this will be the default so you can remove it

bluetech replied:
I'd prefer to keep it so that editor/manual mypy can pick it up.

asottile commented on setup.cfg:
```diff
@@ -61,3 +61,25 @@ ignore =
+# disallow_subclassing_any = True
+# disallow_untyped_calls = True
+# disallow_untyped_decorators = True
+# disallow_untyped_defs = True
```
I would delete all the commented out dead configuration -- one can always look at the configuration reference when adding / removing options

bluetech replied:
I think it's useful as a "TODO"/wishful thinking and for quick experimentation. However, if you don't agree, I'll remove.

asottile replied:
I would delete all the commented out dead configuration

asottile commented on setup.cfg:
```diff
@@ -61,3 +61,25 @@ ignore =
+; implicit_reexport = False
+no_implicit_optional = True
+strict_equality = True
+strict_optional = True
```
this is the default, can remove

nicoddemus commented:
Thanks @bluetech for working on this!

I've changed the title to "WIP" to explicitly state this is not yet prime for merging. 👍

bluetech commented:
I updated the PR to address some of @asottile's comments.

asottile commented on setup.cfg:
```diff
@@ -61,3 +61,25 @@ ignore =
 formats = sdist.tgz,bdist_wheel
+
+[mypy]
+python_version = 3.5
```
I *think* you can drop this since the tests will already fail if invalid syntax is used

bluetech replied:
I'll remove it. I though it'd be nice to have so it already warns about problems in the editor, but CI is still OK I suppose.

asottile commented on src/_pytest/fixtures.py:
```diff
@@ -960,7 +965,8 @@ class FixtureFunctionMarker:
     params = attr.ib(converter=attr.converters.optional(tuple))
     autouse = attr.ib(default=False)
-    ids = attr.ib(default=None, converter=_ensure_immutable_ids)
+    # Ignore type because of https://github.com/python/mypy/issues/6172.
```
hmmm weird, that issue seems to indicate named functions should work, oh well :/

asottile commented on src/_pytest/python.py:
```diff
@@ -1356,7 +1356,10 @@ def write_docstring(tw, doc, indent="    "):
 
 
-class Function(FunctionMixin, nodes.Item, fixtures.FuncargnamesCompatAttr):
+# Ignore type because mypy is confused about Node.repr_failure is defined.
```
is this because the mixin causes a conflict here? I'm not sure I understand the comment :S

bluetech replied:
I think the types match but the way `Node.repr_failure` is defined confused mypy:

https://github.com/pytest-dev/pytest/blob/2180d9ef6d9af45c80fdc89cd64b90df2924c8a7/src/_pytest/nodes.py#L238-L283

But if I replace the assignment with a forwarding method it works out, so I'll do that.

bluetech commented:
Updated again. I've switched to use pre-commit and addresses some of the other comments. Currently some of the tests are failing, but I can't check what causes them exactly ATM - will check later, but I pushed anyway.

asottile commented on .pre-commit-config.yaml:
```diff
@@ -42,6 +43,15 @@ repos:
+        files: ^src/
+    -   id: mypy
+        name: mypy (testing)
+        files: ^testing/
```
oh, I meant you'll probably want one of these for the other dirs you had in the original patch (`extras`, `bench`, etc.) -- it's probably ok to not check testing (I was just using that as an example)

bluetech replied:
I don't think it's worth checking them TBH - they're just some scripts that are probably easier to handle without mypy over your shoulder.

For testing/ I think it's valuable to have type checking, but can be removed later if it's too much to handle.

asottile replied:
the strategy I usually use for testing is `check_untyped_defs` and don't require annotations -- it strikes a nice balance between checled and tedium (generally there's no reason to type check tests because they always run anyway -- and they often run into *intentional* type issues when checking how an api interacts with incorrect types) :man_shrugging:

bluetech replied:
Makes sense. For now, the level of type checking is about what you describe, for the entire project (although `check_untyped_defs` is not activated yet - probably the next logical step).

asottile replied:
sounds good -- thanks for putting so much work into this, this is great :tada:

bluetech commented:
Updated - CI looks happy now. I added `__init__.pyi` files (note: `pyi` not `py`) to workaround the mypy issue without disturbing the tests.

I also changed the mypy pre-commit to use `args: []` so all options are managed exclusively in the setup.cfg.

asottile commented:
@nicoddemus ok with cherry-picking this to master? I'd like to be able to iterate on this there (and not have it regress)

nicoddemus commented:
@asottile it doesn't add any extra dependencies to users of the library or changes any public API, so feel free to go ahead! 👍

RonnyPfannschmidt commented on src/_pytest/_code/code.py:
```diff
@@ -188,11 +189,11 @@ def path(self):
         return self.frame.code.path
 
-    def getlocals(self):
+    @property
```
that one is a breaking change of the api

asottile replied:
I don't think `Code` objects are exposed -- or at least I can't find any cases that they are (?)

blueyed commented on src/_pytest/fixtures.py:
```diff
@@ -31,6 +33,9 @@
 from _pytest.outcomes import TEST_OUTCOME
 
+if False:  # TYPE_CHECKING
+    from typing import Type
```
Why is this only imported conditionally, but not the others?

asottile replied:
it is not available in python 3.5.1 / 3.5.2
